File: examplejava/src/main/java/io/matchmore/sdk/examplejava/MainActivity.java
package io.matchmore.sdk.examplejava;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import io.matchmore.config.SdkConfigTest;
import io.matchmore.sdk.MatchMore;
import io.matchmore.sdk.MatchMoreSdk;
import io.matchmore.sdk.api.models.Publication;
import io.matchmore.sdk.api.models.Subscription;
import kotlin.Unit;


public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Request Location permission
        ActivityCompat.requestPermissions(this, new String[]
                {Manifest.permission.ACCESS_FINE_LOCATION}, 1);

        // Configuration of api key/world id
        if (!MatchMore.isConfigured()) {
            MatchMore.config(this, SdkConfigTest.API_KEY, true);
        }

        // Getting instance. It's static variable. It's possible to have only one instance of matchmore.
        MatchMoreSdk matchMore = MatchMore.getInstance();

        // Creating main device.
        matchMore.startUsingMainDevice(device -> {
            Publication publication = new Publication("Test Topic", 20d, 100000d);
            matchMore.createPublication(publication, createdPublication -> {
                Log.d("JavaExample", createdPublication.getId());
                return Unit.INSTANCE; // `return Unit.INSTANCE;` is important (b/c kotlin vs java lambdas differ in implementation)
            }, e -> {
                Log.d("JavaExample", e.getMessage());
                return Unit.INSTANCE;
            });

            Subscription subscription = new Subscription("Test Topic", 20d, 100000d, "");
            matchMore.createSubscription(subscription, createdSubscription -> {
                Log.d("JavaExample", createdSubscription.getId());
                return Unit.INSTANCE;
            }, e -> {
                Log.d("JavaExample", e.getMessage());
                return Unit.INSTANCE;
            });
            return Unit.INSTANCE;
        }, e -> {
            Log.d("JavaExample", e.getMessage());
            return Unit.INSTANCE;
        });

        // Start getting matches
        matchMore.getMatchMonitor().addOnMatchListener((matches, device) -> {
            Log.d("JavaExample", device.getId());
            return Unit.INSTANCE;
        });


        // Start updating location
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
        } else {
            matchMore.startUpdatingLocation();
        }
    }
}

File: example/src/main/java/io/matchmore/sdk/example/MainActivity.kt
package io.matchmore.sdk.example

import android.Manifest
import android.annotation.SuppressLint
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.widget.Toast
import com.gun0912.tedpermission.PermissionListener
import com.gun0912.tedpermission.TedPermission
import io.matchmore.sdk.MatchMore
import io.matchmore.sdk.MatchMoreSdk
import io.matchmore.sdk.api.models.Publication
import io.matchmore.sdk.api.models.Subscription


class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        MatchMore.instance.apply {
            startUsingMainDevice(
                    { device ->
                        Log.i(TAG, "start using device ${device.name}")

                        val publication = Publication("Test Topic", 1.0, 0.0)
                        publication.properties = hashMapOf("test" to "true")
                        createPublication(publication, { result ->
                            Log.i(TAG, "Publication created ${result.topic}")
                        }, Throwable::printStackTrace)

                        matchMonitor.addOnMatchListener { matches, _ ->
                            Log.i(TAG, "Matches found: ${matches.size}")
                        }
                        createPollingSubscription()
                        checkLocationPermission()
                    }, Throwable::printStackTrace)
        }
    }

    override fun onResume() {
        super.onResume()
        MatchMore.instance.matchMonitor.startPollingMatches()
    }

    override fun onPause() {
        super.onPause()
        MatchMore.instance.matchMonitor.stopPollingMatches()
    }

    override fun onDestroy() {
        super.onDestroy()
        MatchMore.instance.apply {
            stopRanging()
            stopUpdatingLocation()
        }
    }

    private fun MatchMoreSdk.createPollingSubscription() {
        val subscription = Subscription("Test Topic", 1.0, 0.0)
        subscription.selector = "test = 'true'"
        createSubscription(subscription, { result ->
            Log.i(TAG, "Subscription created ${result.topic}")
        }, Throwable::printStackTrace)
    }

    private fun checkLocationPermission() {
        val permissionListener = object : PermissionListener {
            @SuppressLint("MissingPermission")
            override fun onPermissionGranted() {
                MatchMore.instance.apply {
                    startUpdatingLocation()
                    startRanging()
                }
            }

            override fun onPermissionDenied(deniedPermissions: ArrayList<String>) {
                Toast.makeText(this@MainActivity, R.string.if_you_reject, Toast.LENGTH_SHORT).show()
            }
        }
        TedPermission.with(this)
                .setPermissionListener(permissionListener)
                .setDeniedMessage(R.string.if_you_reject)
                .setPermissions(Manifest.permission.ACCESS_FINE_LOCATION)
                .check()
    }

    companion object {
        private const val TAG = "MatchMore"
    }
}
